Return of the Roleplay Context

WolframRavenwolf opened the pull request:
Sorry for the back and forth. I've seen character cards with system prompts as well as various system prompt recommendations on Reddit and realized that for compatibility and ease of use it would be best to separate the format (context) from the contents (system prompt).

Now it's easier to change the system prompt without messing up the format, and character cards can bring their own prompts without messing up the Roleplay preset.

However, now the Roleplay preset needs the proper context template to function properly, like the simple-proxy-for-tavern preset.

***

**Open Questions/Issues:**

- How to link context template and instruct mode preset? At least for Roleplay and simple-proxy-for-tavern, they should always be selected together. **What do you think about this logic:**

  - When a context template is selected, and there's an instruct mode preset with the same name, enable instruct mode and auto-select that preset, and keep it overriding activation regex (as if it were favorited). So a user only has to pick Roleplay or simple-proxy-for-tavern from the context template list, and everything will be set up correctly and permanently for them. If they want to go back to auto-selection based on regex, they'd switch Context Template back to Default or Pygmalion.
  - Optionally, additionally (?): When instruct mode is enabled, if there's a context template with the same name as the active instruct mode preset, auto-select that template.
  - Just what I came up with - open to suggestions/alternatives of how to make this easier to understand and use...

- **IDEA:** Jailbreak prompt for local AI? I've seen character cards with jailbreak prompts that would be useful for local AI, too, so adding that as a variable to the context templates could be helpful?

- **BUG:** When continuing generation (I have "Press "Send" to continue" on), the empty line in front of the output sequence is stripped away. Didn't get a chance to investigate that further, so posting it here as a reminder.

WolframRavenwolf committed:
Return of the Roleplay Context

Cohee1207 commented:
> When a context template is selected, and there's an instruct mode preset with the same name, enable instruct mode and auto-select that preset

Ok.

> Optionally, additionally (?): When instruct mode is enabled, if there's a context template with the same name as the active instruct mode preset, auto-select that template.

Ok, but when? On preset selection, including auto-select?

> IDEA: Jailbreak prompt for local AI? I've seen character cards with jailbreak prompts that would be useful for local AI, too, so adding that as a variable to the context templates could be helpful.

Where would that go prompt-wise? Story string too? Cause params like scenario, system, etc. are only supported there. Otherwise, it needs to do chat injections as for the author's note. The second case is not very transparent for the users. Also, rue the day when local model jailbreaks became a necessity.  

> BUG: When continuing generation (I have "Press "Send" to continue" on), the empty line in front of the output sequence is stripped away. Didn't get a chance to investigate that further, so posting it here as a reminder.

Is this recent? May be something with the recent CFG additions from #968, but I'm not sure, I haven't tested it with instruct mode yet.

WolframRavenwolf commented:
Does "Ok" mean "Ok, you can do it" or "Ok, I'll do it"? 😉 Whoever does it, we should avoid duplication of effort, so let me know if it's something you're planning to do yourself or if you want me to do that.

The "When" part of the optional suggestion is what's bugging me, too. We don't want a loop here, and we also don't want unexpected changes (user turning on instruct mode while Roleplay/proxy preset is selected, thus changing the template above, possibly without the user noticing). I have no satisfying answer for this yet.

So much for now, will be out today, so can only check back later.

Cohee1207 commented:
It was an acknowledgment of the proposition. I wasn't going to do it myself yet, you can do it. I need to make some other improvements to the story string templates (unrelated to instruct mode, so should be little to no conflicts).

Cohee1207 commented:
Also as much as I love the cool emoji, it would need to be removed from the preset name.
1. It blocks the matching of presets by names.
2. It causes a duplication if updated via ST, because of names mismatch.

WolframRavenwolf committed:
auto select presets

WolframRavenwolf commented:
**Auto Select Presets:**

I've found a way to handle auto-selecting presets that is both user-friendly and logical:

- 1st Goal: Prevent broken configurations, like a context template that needs an instruct preset to work, and vice versa.
- 2nd Goal: Make enabling the Roleplay or simple-proxy-for-tavern presets a one-click thing for ease of use.
- 3rd Goal: Don't interfere with regex activation of other presets.
- Preset association is based on preset names, if they have the same name, they are associated. Currently, that's only Roleplay and simple-proxy-for-tavern.
- When user selects a context template that has an associated instruct preset, that instruct preset is selected. If instruct mode disabled, it is enabled.
- When user selects an instruct template that has an associated context template, that context template is selected. If instruct mode disabled, it is enabled.
- When user enables instruct mode, if an instruct template is selected that has an associated context template, that context template is selected.
- For presets without associated presets, follow the normal selection rules, e. g. activation regex or favoriting.

***

OK, if it's problematic, I'll remove the wizard emoji. 😢

WolframRavenwolf committed:
byebye wizard

WolframRavenwolf commented:
**Byebye Wizard:**

🧙 *poof*

***

I still want to improve preset selection logic further: Right now, changing from Roleplay or Proxy instruct preset to another won't change the context template, but these presets should be used with the Default context template. So for ease of use, it would be best if selecting an instruct template that doesn't have an associated context template would fall back to Default. I'll add that, so don't merge yet, please.

Cohee1207 commented on public/instruct/! Roleplay.json:
```diff
@@ -1,6 +1,6 @@
 {
```
Don't forget to remove `! ` from the file name as well. ST writes preset based on "name" property in the preset body, so it should match

WolframRavenwolf replied:
Done.

Cohee1207 commented on default/settings.json:
```diff
@@ -135,7 +135,7 @@
         "instruct": {
             "enabled": false,
             "preset": "🧙 Roleplay",
```
Need to remove it from here as well.

WolframRavenwolf replied:
Done.

Cohee1207 commented on public/scripts/instruct-mode.js:
```diff
@@ -82,20 +123,17 @@ export function autoSelectInstructPreset(modelId) {
-        // If activation regex is set, check if it matches the model id
-        if (preset.activation_regex) {
+        // If context template matches the preset name
+        if (power_user.context.preset === preset.name) {
```
I think you should loop over instruct_presets twice.
Because it will stop on the first valid regex instead of checking all presets for a match with the current context preset. Unless I'm missing something.

WolframRavenwolf replied:
Yes, you're right, thanks for pointing that out. Fixed.

WolframRavenwolf committed:
improved preset selection logic

WolframRavenwolf committed:
renamed roleplay instruct preset

WolframRavenwolf commented:
**improved preset selection logic:**

- Fixed and improved, I think that's it now!

**renamed roleplay instruct preset:**

- Removed special characters for tweaked sorting order.

Cohee1207 commented:
There's still a bug left with the newline on the last prompt line getting trimmed, right?

Cohee1207 commented:
I checked your code, works fine for me. Just let me know how should we handle the bug (if it still exists) before merging.

WolframRavenwolf commented:
The bug still exists. When I click send to continue generation, the new prompt is no longer using the Last Sequence, but the Output Sequence. And it's missing the newline that I have in there in front of the string. Probably a .trim() that's doing that?

Cohee1207 commented:
Okay, I'll take a look soon.

WolframRavenwolf commented:
@Cohee1207 Found where the newline gets trimmed:

[SillyTavern/public/script.js at staging · SillyTavern/SillyTavern](https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L2609)

After that, there's also the issue that the Output Sequence gets added instead of the Last Sequence. That's stored in generatedPromtCache.

Cohee1207 commented:
Got them both... I think.

Cohee1207 commented:
@StefanDanielSchwarz one user brought up today that our llama2 prompt template is completely broken. I could verify that it doesn't produce replies. Since you're pretty active in the local models community, any ideas on how to fix that with prompt + instruct templates? llama2 is notoriously and unnecessarily difficult to get right. I got something slightly working, but it tends to start replies with the output sequence.
![image](https://github.com/SillyTavern/SillyTavern/assets/18619528/55c27bd2-f984-4ea8-a0bb-e6ac977a90f9)

WolframRavenwolf commented:
Challenge accepted!

[Llama 2 chat instruct preset fixes by StefanDanielSchwarz · Pull Request #1012 · SillyTavern/SillyTavern](https://github.com/SillyTavern/SillyTavern/pull/1012)
